### PR TITLE
making the conditional rules in tests repeatable

### LINKED ISSUE #1669: @ConditionalRun needs to be repeatable
`ProjectMessageTest` needs multiple conditions (check whether Git, Mercurial and Subversion repositories are working). In order to achieve this in elegant way, `@ConditionalRun` annotation needs to be repeatable

### PR DESCRIPTION
**Pull request by tulinkry**
fixes #1669

### CODE REVIEW

**Comment by tulinkry** [comment]
One downside of this is that there will always be 2 skipped tests in junit (which indicates correct behaviour).

**Comment by vladak** [comment]
While you are at it could you use the repeatable condition in `ProjectMessageTest` ?

File: test/org/opensolaris/opengrok/condition/ConditionalRunRule.java [hunk 1]
```diff
@@ -64,34 +68,70 @@ public Statement apply(Statement aStatement, Description aDescription) {
...
+            throw new RuntimeException(ex);
+        }
     }
 
     private static RunCondition getIgnoreConditionOnMethod(Description aDescription) {
-        ConditionalRun annotation = aDescription.getAnnotation(ConditionalRun.class);
-        return new IgnoreConditionCreator(aDescription.getTestClass(), annotation).create();
+        try {
+            // this is possible because test methods must not have any argument
+            ConditionalRun[] annotation = aDescription.getTestClass().getMethod(aDescription.getMethodName()).getAnnotationsByType(ConditionalRun.class);
```
Comment (vladak) [inline test/org/opensolaris/opengrok/condition/ConditionalRunRule.java]: annotations ?

Comment (tulinkry) [reply test/org/opensolaris/opengrok/condition/ConditionalRunRule.java]: fixed

**Comment by tulinkry** [comment]
> While you are at it could you use the repeatable condition in ProjectMessageTest ?

I wanted to do this in a different PR.

**Comment by tulinkry** [comment]
@vladak ?

File: test/org/opensolaris/opengrok/condition/RunningRepeatableConditionTest.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    @ConditionalRun(condition = TrueRunCondition.class)
+    @ConditionalRun(condition = FalseRunCondition.class)
+    @ConditionalRun(condition = TrueRunCondition.class)
+    @Test
+    public void testSkippedTest() throws NoSuchMethodException {
+        Assert.assertTrue("This test must be skipped", false);
+    }
+
+    public static class TrueRunCondition implements RunCondition {
```
Comment (vladak) [inline test/org/opensolaris/opengrok/condition/RunningRepeatableConditionTest.java]: somehow it seems strange to me having the `{True,False}RunCondition` being shared like this. Why not have them in separate file ?

**Comment by vladak** [comment]
Separate PR for the actual use in tests is fine, as well as separate commit in this PR.

File: test/org/opensolaris/opengrok/condition/ConditionalRunRule.java [hunk 3]
```diff
@@ -64,34 +68,70 @@ public Statement apply(Statement aStatement, Description aDescription) {
...
+            }
+            return true;
+        }
+    }
+
+    protected static class IgnoreConditionCreator {
 
         private final Class<?> mTestClass;
-        private final Class<? extends RunCondition> conditionType;
+        private final List<Class<? extends RunCondition>> conditionType;
```
Comment (vladak) [inline test/org/opensolaris/opengrok/condition/ConditionalRunRule.java]: rename to signify it is now holding more conditions ?

File: test/org/opensolaris/opengrok/condition/ConditionalRunRule.java [hunk 4]
```diff
@@ -64,34 +68,70 @@ public Statement apply(Statement aStatement, Description aDescription) {
...
 
-        IgnoreConditionCreator(Class<?> aTestClass, ConditionalRun annotation) {
+        public IgnoreConditionCreator(Class<?> aTestClass, ConditionalRun[] annotation) {
             this.mTestClass = aTestClass;
-            this.conditionType = annotation.condition();
+            this.conditionType = new ArrayList<>(annotation.length);
+            for (int i = 0; i < annotation.length; i++) {
```
Comment (vladak) [inline test/org/opensolaris/opengrok/condition/ConditionalRunRule.java]: would `addAll()` work here ?

File: test/org/opensolaris/opengrok/condition/ConditionalRun.java [hunk 5]
```diff
@@ -18,19 +18,25 @@
...
+import java.lang.annotation.Repeatable;
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
 
+/**
+ * Repeatable annotation for automatic test skipping.
+ */
 @Retention(RetentionPolicy.RUNTIME)
 @Target(value = {ElementType.METHOD, ElementType.TYPE})
```
Comment (vladak) [inline test/org/opensolaris/opengrok/condition/ConditionalRun.java]: just wondering if expanding this with `PACKAGE` would make sense

File: test/org/opensolaris/opengrok/condition/ConditionalRun.java [hunk 6]
```diff
@@ -18,19 +18,25 @@
...
+@Repeatable(ConditionalRunRepeatable.class)
 @Documented
 public @interface ConditionalRun {
```
Comment (vladak) [inline test/org/opensolaris/opengrok/condition/ConditionalRun.java]: use [`Inherited`](https://docs.oracle.com/javase/8/docs/api/java/lang/annotation/Inherited.html) ?

File: test/org/opensolaris/opengrok/condition/ConditionalRunRepeatable.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * Annotation allowing the {@link ConditionalRun} to be repeatable annotation.
+ *
+ * @author Krystof Tulinger
+ */
+@Retention(value = RUNTIME)
+@Documented
+@Target(value = {ElementType.METHOD, ElementType.TYPE})
+public @interface ConditionalRunRepeatable {
+
+    ConditionalRun[] value();
```
Comment (vladak) [inline test/org/opensolaris/opengrok/condition/ConditionalRunRepeatable.java]: I am slightly confused by the `value` used in this file. Does it have any special significance to name the array so ?

Comment (vladak) [reply test/org/opensolaris/opengrok/condition/ConditionalRunRepeatable.java]: Oh, this seems to be the rule for containing annotation type as per https://docs.oracle.com/javase/tutorial/java/annotations/repeating.html

Comment (tulinkry) [reply test/org/opensolaris/opengrok/condition/ConditionalRunRepeatable.java]: yes it must be like this

File: test/org/opensolaris/opengrok/condition/ConditionalRunRule.java (hunk 3 above)
Comment (tulinkry) [reply test/org/opensolaris/opengrok/condition/ConditionalRunRule.java]: fixed

File: test/org/opensolaris/opengrok/condition/ConditionalRunRule.java (hunk 4 above)
Comment (tulinkry) [reply test/org/opensolaris/opengrok/condition/ConditionalRunRule.java]: no cause you still need to call annotation[i].condoition()

File: test/org/opensolaris/opengrok/condition/ConditionalRun.java (hunk 5 above)
Comment (tulinkry) [reply test/org/opensolaris/opengrok/condition/ConditionalRun.java]: i have no idea how this should behave but we can experiment with it

File: test/org/opensolaris/opengrok/condition/ConditionalRun.java (hunk 6 above)
Comment (tulinkry) [reply test/org/opensolaris/opengrok/condition/ConditionalRun.java]: added

File: test/org/opensolaris/opengrok/condition/RunningRepeatableConditionTest.java (hunk 2 above)
Comment (tulinkry) [reply test/org/opensolaris/opengrok/condition/RunningRepeatableConditionTest.java]: I made them protected. There's no need to use them anywhere else.

**Comment by tarzanek** [comment]
let's merge

**Pull request merged**